src/proxy_protocol/mod.rs

```diff
@@ -0,0 +1,274 @@
+            return Ok((stream, None));
+        }
+
+        buffer.clear();
+    };
+
+    match header {
+        HeaderResult::V1(Ok(_header)) => {
+            // V1 PROXY header detected, which is not supported
+            Ok((stream, None))
+        }
+        HeaderResult::V2(Ok(header)) => {
+            let client_address = match header.addresses {
+                v2::Addresses::IPv4(ip) => IpAddr::V4(ip.source_address),
+                v2::Addresses::IPv6(ip) => IpAddr::V6(ip.source_address),
+                v2::Addresses::Unix(_unix) => {
+                    // Unix socket addresses are not supported
+                    return Ok((stream, None));
+                }
+                v2::Addresses::Unspecified => {
+                    // V2 PROXY header addresses "Unspecified"
+                    return Ok((stream, None));
+                }
+            };
+
+            // Remove header length in bytes from the stream
+            let header_len = header.len();
+            stream
+                .read_exact(&mut buffer.filled_mut()[..header_len])
+                .await?;
+
+            Ok((stream, Some(client_address)))
+        }
+        HeaderResult::V1(Err(_error)) => {
+            // No valid V1 PROXY header received
+            Ok((stream, None))
+        }
+        HeaderResult::V2(Err(_error)) => {
+            // No valid V2 PROXY header received
+            Ok((stream, None))
+        }
+    }
+}
+
+/// Middleware for adding client IP address to the request `forwarded` header.
+/// see spec: <https://www.rfc-editor.org/rfc/rfc7239#section-5.2>
+#[derive(Debug)]
+pub struct ForwardClientIp<S> {
+    inner: S,
+    client_address_opt: Option<IpAddr>,
+}
+
+impl<S> Service<Request<hyper::Body>> for ForwardClientIp<S>
+where
+    S: Service<Request<hyper::Body>>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: Request<hyper::Body>) -> Self::Future {
+        let forwarded_string = match self.client_address_opt {
+            Some(ip) => match ip {
+                IpAddr::V4(ipv4) => {
+                    format!("for={}", ipv4)
+                }
+                IpAddr::V6(ipv6) => {
+                    format!("for=\"[{}]\"", ipv6)
+                }
+            },
+            None => "for=unknown".to_string(),
+        };
+
+        if let Ok(header_value) = HeaderValue::from_str(&forwarded_string) {
+            req.headers_mut().insert("Forwarded", header_value);
+        }
+
+        self.inner.call(req)
+    }
+}
+
+#[derive(Clone)]
+pub struct ProxyProtocolAcceptor<A> {
+    inner: A,
+    parsing_timeout: Duration,
+}
+
+impl<A> ProxyProtocolAcceptor<A> {
+    pub(crate) fn new(inner: A) -> Self {
+        #[cfg(not(test))]
```

Comment:
Seems bad to have this here, it should instead be passed in a constructor/config in the test itself.